src/opam.nix

```diff
@@ -480,4 +482,120 @@ in rec {
+      nameFromURL = url: sep:
+        let
+          components = splitString "/" url;
+          filename = last components;
+          name = head (splitString sep filename);
+        in name;
+      defToSrc = { version, ... }@pkgdef:
+        let
+          inherit (getUrl bootstrapPackages pkgdef) src;
+          name = let n = nameFromURL pkgdef.dev-repo "."; in
+                 # rename dune so it doesn't clash with dune file in duniverse
+                 if n == "dune" then "_dune" else n;
+        in
+        # filter out pkgs without dev-repos
+        if pkgdef ? dev-repo then { inherit name version src; } else { };
+      # remove filterPkgs
+      filteredDefs = removeAttrs defs filterPkgs;
+      srcs = mapAttrsToList (pkgName: def: defToSrc def) filteredDefs;
+      # remove empty elements from pkgs without dev-repos
+      cleanedSrcs = remove { } srcs;
+    in cleanedSrcs;
+
+  deduplicateSrcs = srcs:
+    # This is O(n^2). We could try and improve this by sorting the list on name. But n is small.
+    let op = srcs: newSrc:
+      # Find if two packages come from the same dev-repo.
+      # Note we are assuming no dev-repos will have different names here, but we also assume
+      # this later when we will symlink in the duniverse directory based on this name.
+      let duplicateSrc = findSingle (src: src.name == newSrc.name) null "multiple" srcs; in
+      # Multiple duplicates should never be found as we deduplicate on every new element.
+      assert duplicateSrc != "multiple";
+      if duplicateSrc == null then srcs ++ [ newSrc ]
+      # > If packages from the same repo were resolved to different URLs, we need to pick
+      # > a single one. Here we decided to go with the one associated with the package
+      # > that has the higher version. We need a better long term solution as this won't
+      # > play nicely with pins for instance.
+      # > The best solution here would be to use source trimming, so we can pull each individual
+      # > package to its own directory and strip out all the unrelated source code but we would
+      # > need dune to provide that feature.
+      # See [opam-monorepo](https://github.com/tarides/opam-monorepo/blob/9262e7f71d749520b7e046fbd90a4732a43866e9/lib/duniverse.ml#L143-L157)
+      else if duplicateSrc.version >= newSrc.version then srcs
+      else (remove duplicateSrc srcs) ++ [ newSrc ];
+    in foldl' op [ ] srcs;
+
+  mkMonorepo = srcs:
+    let
+      # derivation that fetches the source
+      mkSrc = { name, version, src }:
+        bootstrapPackages.pkgsBuildBuild.stdenv.mkDerivation ({
+          inherit name version src;
+          phases = [ "unpackPhase" "installPhase" ];
+          installPhase = ''
+            mkdir $out
+            cp -R . $out
+          '';
+        });
+    in
+    listToAttrs (map (src: nameValuePair src.name (mkSrc src)) srcs);
+
+  queryToMonorepo = { repos ? [ opamOverlays opamRepository ], resolveArgs ? { }
```

Comment:
Any reason to not use `opamRepositoryModuloOverlay` here?

Reply:
Nope! Good catch